### DISCUSSION TITLE
STM32F4 dual cdc problem

### DISCUSSION THREAD

**Discussion by TOMASHEVSKIYIVAN** [hathach/tinyusb#2432] [Q&A]
Hello everyone! I’m trying to write a program for the STM32F4 to run 2 CDC on 1 interface. I’ve read that this microcontroller is said to have not enought endpoints, but how can this be possible if the NUCLEO board uses the STM32F103 microcontroller in the ST-Link part and when you connect the NUCLEO to a PC, it identifies the ST-Link and the MSD. I don’t understand how the STM32F4 can’t support 2 CDCs, so can you please explain in more detail?

**Comment by TOMASHEVSKIYIVAN** [comment]
I made firmware to PC recognize 2 CDC, however couldn't make device to sand information on second CDC

**Comment by Unturned3** [comment]
I haven't been able to get the dual CDC example to work on the STM32F411. According to its datasheet, the F411 only has 4 endpoints, but dual CDC needs 6. On the [STM32F103 datasheet (page 19)](https://www.st.com/resource/en/datasheet/stm32f103rb.pdf), it mentions MCU has "software-configurable endpoint setting" instead of a fixed number. Does this mean that it can somehow "create" USB endpoints on-demand?

I'm not too familiar with this topic in general. Perhaps @hathach can comment on this?

**Comment by HiFiPhile** [comment]
>   it mentions MCU has "software-configurable endpoint setting" instead of a fixed number. Does this mean that it can somehow "create" USB endpoints on-demand

It means endpoint type (bulk,iso,int) and buffer size are configurable.

STM32F103 use fsdev driver and has 8 endpoints.

**Comment by hathach** [comment]
- 1 CDC need 2 IN endpoint + 1 OUT endpoint
- F411 only has 3 IN endpionts + 3 OUT endpoints

there is simply not enough IN endpoints for F411.
